provider/aws: ElastiCache (Redis) snapshot window and retention limits

stack72 opened the pull request:
This fixes #3075

```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=ElasticacheCluster_snapshots' 2>~/tf.log
go generate ./...
TF_ACC=1 go test ./builtin/providers/aws -v -run=ElasticacheCluster_snapshots -timeout 90m
=== RUN   TestAccAWSElasticacheCluster_snapshots
--- PASS: TestAccAWSElasticacheCluster_snapshots (366.93s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	366.946s
```

Also tested the ValidateFunc:
```
terraform [aws-elasticache-cluster-snapshoting●] % make testacc TEST=./builtin/providers/aws TESTARGS='-run=ElasticacheCluster_snapshots' 2>~/tf.log
go generate ./...
TF_ACC=1 go test ./builtin/providers/aws -v -run=ElasticacheCluster_snapshots -timeout 90m
=== RUN   TestAccAWSElasticacheCluster_snapshots
--- FAIL: TestAccAWSElasticacheCluster_snapshots (0.11s)
	testing.go:137: Step 0 error: Configuration is invalid.

		Warnings: []string(nil)

		Errors: []string{"aws_elasticache_cluster.bar: snapshot retention limit cannot be more than 35 days"}
FAIL
```
Creates the values as can be seen in the console:

<img width="977" alt="screen shot 2015-10-30 at 23 57 53" src="https://cloud.githubusercontent.com/assets/227823/10860577/1f53ae2a-7f62-11e5-8fcc-8b2262cbfd1b.png">

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster.go:
```diff
@@ -344,6 +372,15 @@ func resourceAwsElasticacheClusterUpdate(d *schema.ResourceData, meta interface{
 	}
 
+	if d.HasChange("snapshot_window") {
+		req.EngineVersion = aws.String(d.Get("snapshot_window").(string))
```
Probably not `EngineVersion` :smile:

stack72 replied:
D'OH!

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster.go:
```diff
@@ -344,6 +372,15 @@ func resourceAwsElasticacheClusterUpdate(d *schema.ResourceData, meta interface{
+	}
+
+	if d.HasChange("snapshot_retention_limit") {
+		req.NumCacheNodes = aws.Int64(int64(d.Get("snapshot_retention_limit").(int)))
```
Probably not `NumCacheNodes` :smile:

stack72 replied:
D'OH!

stack72 commented:
@catsby ok, sorry about that. these silly copy paste errors have been fixed now

catsby commented:
Thanks @stack72 – some where things here... I thought you'd need to specify `Computed: true` for these attributes, but then I found you did not... so I dug in and found that SnapshotRetentionLimit and SnapshotWindow are nil in the API response... not sure why. 

I did try to update the `SnapshotWindow` attribute and got a weird error:

```console
* aws_elasticache_cluster.test: [WARN] Error updating ElastiCache cluster (test), error: InvalidParameterCombination: Instance type does not support snapshotting. Snapshot window parameter should not be specified.
	status code: 400, request id: 7e2dcf24-81ab-11e5-8466-199fd0904dd9
```

Not sure what that's about, the debug output of the update didn't mention instance type..

Plan:

```console
~ aws_elasticache_cluster.test
    snapshot_window: "Wed:09:00-Wed:09:30" => "05:00-09:00"
```

Log: 
```
[DEBUG]: 2015/11/02 15:48:47 [DEBUG] Modifying ElastiCache Cluster (test), opts:
[DEBUG]: {
[DEBUG]:   ApplyImmediately: false,
[DEBUG]:   CacheClusterId: "test",
[DEBUG]:   SnapshotWindow: "05:00-09:00"
[DEBUG]: }
```

Do you get this behavior too?

stack72 commented:
@catsby I added an acceptance test to show these in motion and to make sure that they are set as expected and both worked. I was able to control my instance based on this as well

I think you are specifying a Memcached cluster?

catsby commented:
Nah, I'm using Redis:

```hcl
provider "aws" {
  region = "us-west-2"
}
resource "aws_vpc" "default" {
  cidr_block = "10.250.0.0/16"
  tags {
    Name = "ec-sub-test"
  }
}

resource "aws_subnet" "private" {
  vpc_id = "${aws_vpc.default.id}"
  cidr_block = "10.250.3.0/24"
}

resource "aws_subnet" "private2" {
  vpc_id = "${aws_vpc.default.id}"
  cidr_block = "10.250.2.0/24"
}

resource "aws_elasticache_subnet_group" "test" {
  name        = "test"
  description = "Test Redis"
  subnet_ids = [
    "${aws_subnet.private.*.id}", 
    "${aws_subnet.private2.id}",
  ]
}

resource "aws_elasticache_cluster" "test" {
  cluster_id = "test"
  subnet_group_name = "${aws_elasticache_subnet_group.test.name}"
  num_cache_nodes = 1
  node_type = "cache.t2.micro"
  engine = "redis"
  port = 6379

  snapshot_retention_limit = 1
  snapshot_window = "05:00-09:00"
}
```

The **code** seems to work, but adding debug statements in the `read` shows those attributes are nil from the API... at least for me..

stack72 commented:
I'll have another look at the debugging of this in the morning and print some info

Thanks for checking this

stack72 commented:
@catsby ok, I just added an extra test to make sure that updating works as expected. Also added some debug logic to the file to make sure that it writes INFO messages to the tf.log when running tests

I definitely get results on my side and I am using this logic in my code from this morning as a check

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster_test.go:
```diff
@@ -149,6 +204,76 @@ resource "aws_elasticache_cluster" "bar" {
     parameter_group_name = "default.memcached1.4"
     security_group_names = ["${aws_elasticache_security_group.bar.name}"]
+    snapshot_window = "05:00-09:00"
+    snapshot_retention_limit = 3
```
Engine `memcached` does not support snapshotting, so the `TestAccAWSElasticacheCluster_basic` test now fails with this addition :/

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster.go:
```diff
@@ -261,6 +287,10 @@ func resourceAwsElasticacheClusterRead(d *schema.ResourceData, meta interface{})
 		d.Set("security_group_ids", c.SecurityGroups)
 		d.Set("parameter_group_name", c.CacheParameterGroup)
 		d.Set("maintenance_window", c.PreferredMaintenanceWindow)
+		log.Printf("[INFO] Found %s as the Snapshow Window", *c.SnapshotWindow)
```
we probably don't want to leave these in, but it's bizarre.. these lines panic for me if I don't specify a snapshot window. Does this config work for you? 

```hcl
provider "aws" {
  region = "us-west-2"
}
resource "aws_vpc" "default" {
  cidr_block = "10.250.0.0/16"
  tags {
    Name = "ec-sub-test"
  }
}

resource "aws_subnet" "private" {
  vpc_id = "${aws_vpc.default.id}"
  cidr_block = "10.250.3.0/24"
}

resource "aws_subnet" "private2" {
  vpc_id = "${aws_vpc.default.id}"
  cidr_block = "10.250.2.0/24"
}

resource "aws_elasticache_subnet_group" "test" {
  name        = "test"
  description = "Test Redis"
  subnet_ids = [
    "${aws_subnet.private.*.id}", 
    "${aws_subnet.private2.id}",
  ]
}

resource "aws_elasticache_cluster" "test" {
  cluster_id = "snapshot-tests"
  subnet_group_name = "${aws_elasticache_subnet_group.test.name}"
  num_cache_nodes = 1
  node_type = "cache.t2.micro"
  engine = "redis"
  port = 6379

  #snapshot_retention_limit = 1
  #snapshot_window = "05:00-09:00"
}
```

catsby commented:
> I definitely get results on my side and I am using this logic in my code from this morning as a check

Do you get results if those attributes are omitted in the config?

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster.go:
```diff
@@ -344,6 +374,14 @@ func resourceAwsElasticacheClusterUpdate(d *schema.ResourceData, meta interface{
 		requestUpdate = true
 	}
 
+	if d.HasChange("snapshot_window") {
```
`snapshot_window` and `snapshot_retention_limit` are not toggling the `requestUpdate` bit, so they're never actually being sent in any `update` call.

catsby commented on website/source/docs/providers/aws/r/elasticache_cluster.html.markdown:
```diff
@@ -73,6 +73,15 @@ names to associate with this cache cluster
+SnapshotRetentionLimit to 5, then a snapshot that was taken today will be retained for 5 days 
+before being deleted. If the value of SnapshotRetentionLimit is set to zero (0), backups are turned off. 
+Can only be used for the Redis engine.
+
```
Another gem: snapshotting is not enabled for `t2` instance types, like the ones I've been testing with (explaining much :laughing: )

> Redis backup/restore is not supported for t2 instances.

- http://docs.aws.amazon.com/fr_fr/AmazonElastiCache/latest/APIReference/API_CreateCacheCluster.html

stack72 commented:
ok, @catsby all is working as expected now :)

I was able to code out the cache.t2 sizing as follows:

```
if !strings.Contains(d.Get("node_type").(string), "cache.t2") {
		if v, ok := d.GetOk("snapshot_retention_limit"); ok {
			req.SnapshotRetentionLimit = aws.Int64(int64(v.(int)))
		}

		if v, ok := d.GetOk("snapshot_window"); ok {
			req.SnapshotWindow = aws.String(v.(string))
		}
	}
```

I was also able to get the updates working (was missing requestUpdate=true):

```
if d.HasChange("snapshot_window") {
			req.SnapshotWindow = aws.String(d.Get("snapshot_window").(string))
			requestUpdate = true
		}

		if d.HasChange("snapshot_retention_limit") {
			req.SnapshotRetentionLimit = aws.Int64(int64(d.Get("snapshot_retention_limit").(int)))
			requestUpdate = true
		}
```

also made sure not to cause panics in the Read event:

```
if c.SnapshotWindow != nil {
			d.Set("snapshot_window", c.SnapshotWindow)
		}
		if c.SnapshotRetentionLimit != nil {
			d.Set("snapshot_retention_limit", c.SnapshotRetentionLimit)
		}
```

The acceptance tests all work as expected now.

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster.go:
```diff
@@ -187,6 +205,16 @@ func resourceAwsElasticacheClusterCreate(d *schema.ResourceData, meta interface{
 		req.CacheParameterGroupName = aws.String(v.(string))
 	}
 
+	if !strings.Contains(d.Get("node_type").(string), "cache.t2") {
```
I don't believe this is the approach we should take. We should document that snapshotting does not work with `cache.t2`'s , and we should bubble up the error if a user uses t2's anyway. 

Only applying these params in this block would (especially without any message) would be misleading the user, as without an error, anything in their config that doesn't throw an error in applying will be written to state, so from Terraforms view, they are applied. But in reality, they are not.

stack72 replied:
So you thibk I should just log that snapshotting isn't available if the user specifies t2 and then just let the values be set? Or should I specifically check for snapahotting in conjunction with t2 (using a check like above) and throw an error?

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster.go:
```diff
@@ -261,6 +289,12 @@ func resourceAwsElasticacheClusterRead(d *schema.ResourceData, meta interface{})
 		d.Set("security_group_ids", c.SecurityGroups)
 		d.Set("parameter_group_name", c.CacheParameterGroup)
 		d.Set("maintenance_window", c.PreferredMaintenanceWindow)
+		if c.SnapshotWindow != nil {
```
`d.Set` can handle `nil`s, so this guard isn't necessary. It would only be necessary if you're wanting to log our their value

catsby commented:
Left a few additional notes, and we still need to document the snapshot limitation regarding t2s, otherwise looking good :smile:

catsby commented:
> The acceptance tests all work as expected now EXCEPT the vpc test which throws a panic. I just tried this on master and get the same error I'm afraid

Yup, that's my fault :frowning: it gets fixed in https://github.com/hashicorp/terraform/pull/3777

stack72 commented:
ok, @catsby, I have documented the t2 issue and now let the error bubble up to the user. I now get the following:

```
	* aws_elasticache_cluster.bar: Error creating Elasticache: InvalidParameterCombination: Instance type does not support snapshotting. Snapshot window parameter should not be specified.
			status code: 400, request id: 7b21df4b-8477-11e5-be36-5514cd0a8d29
```

This goes hand in hand with the NOTE added to the docs

catsby commented on builtin/providers/aws/resource_aws_elasticache_cluster_test.go:
```diff
@@ -152,6 +191,75 @@ resource "aws_elasticache_cluster" "bar" {
+resource "aws_elasticache_cluster" "bar" {
+    cluster_id = "tf-test-%03d"
+    engine = "redis"
+    node_type = "cache.t2.small"
```
this needs to be changed to `cache.m1.small`, or `snapshot_window` and `snapshot_retention_limit` need to be removed. The test that uses this config doesn't pass

catsby replied:
Went ahead and made this change myself, thanks!

catsby commented:
Thanks!
